### ISSUE TITLE
Inject Gaussian noise into the recurrent latent to force real work per refine pass

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#219]
> **⚠ Premise corrected — read the pinned comment before the body below.**
> The motivation here ("the block has learned a lazy identity / idles") is **measurably false**:
> the block moves and changes ~1% of predicted tokens; it just gains nothing. The experimental
> design is still sound, the reasoning for it is not. Gated behind #227.

## Hypothesis

Corrupting the recurrent latent `z` with Gaussian noise during training forces the shared
refine block to do real work on every pass, instead of settling into a near-identity
fixed point. Read out as **depth sensitivity**, not just final CE.

## Why — the observation that motivates it

Measured on the live `run_20260813_214725` checkpoint at 90% (single position, `the capital
of france is`):

```
depth   |z − z_d2|      |logits − logits_d2|   argmax
1       5.42e-01        1.05e-01               7993
4       2.53e-01        5.27e-02               7993
8       2.10e-01        3.04e-02               7993
```

Depth moves the latent an order of magnitude more than it moves the logits. Two structural
reasons this is expected rather than paradoxical, both worth stating because they bound
what any fix can achieve:

- **`out_norm` (RMSNorm) deletes the radial component.** `logits = out_norm(z) @ Eᵀ`, so
  any refinement that only grows or shrinks `z` is invisible to the readout by construction.
- **Softmax deletes the common mode.** Motion along a direction with near-equal dot product
  against all 50k embedding rows shifts every logit together and changes nothing. LM
  embedding spaces are strongly anisotropic — a dominant common direction carries most of
  the variance (Ethayarajh 2019; Timkey & van Schijndel 2021, "rogue dimensions") — and
  that direction is exactly the one softmax ignores.

So a large `|Δz|` with a small `|Δlogits|` is consistent with the refine block spending its
compute in the readout's blind spot.

### But the effect is real and *growing*

Byte-identical completions across the 8 logbook prompts, from the three GPU transcript
entries already on disk (`runs/run_20260813_214725/transcripts/`):

| identical pairs (of 8) | step 14,400 (47%) | 24,384 (80%) | 27,456 (90%) |
|---|---|---|---|
| d1 vs d2 | 4/8 | 4/8 | 4/8 |
| d2 vs d4 | 7/8 | 6/8 | 5/8 |
| d4 vs d8 | **8/8** | 7/8 | 6/8 |

At 47% depth 4 and depth 8 produced the *same text on every prompt* — refinement fully
saturated past 4 passes. By 90% they differ on 2 of 8. **Depth sensitivity emerges on its
own over training.** This is why the framing here is "accelerant", not "rescue": the block
is not broken, it is slow. An earlier read of mine that called it a "lazy identity" was
based on n=1 and is withdrawn.

## Design

Add `state_noise: float = 0.0` to `CausalRefiner`. Two variants, and the difference is the
whole point of the experiment:

- **`init`** — perturb `z` once, after the encoder, before the refine loop.
- **`per_pass`** — perturb `z` before every refine iteration inside the loop.

Train-time only; inference runs clean.

**No auxiliary reconstruction loss.** The obvious version of this idea is "add noise, then
add a loss term that reconstructs the original `z`". Don't. That target is the model's own
hidden state, and regressing a network onto its own representation has a well-known
degenerate solution — collapse to a constant. It is the BYOL / SimSiam problem, and those
needed a momentum encoder or a stop-grad predictor to avoid it. Our CE loss blocks *total*
collapse, but partial collapse (`z` going low-rank) would be invisible except as slightly
worse val CE. **CE already is the reconstruction target**: recovering the state is the only
way to predict the token. Zero new loss terms, zero new hyperparameters beyond σ, zero
collapse risk.

## Prior art and the novelty line

Geiping et al. 2025 (depth-recurrent "Huginn", 3.5B — *Scaling up Test-Time Compute with
Latent Reasoning: A Recurrent Depth Approach*) initialise the recurrent state from Gaussian
noise rather than a deterministic embedding, and randomise the recurrence count during
training. We already do the second half via `MAX_STEPS_LIMIT=8`.

So:
- **`init` is settled.** Running it confirms a known result in our setting — PR is the
  record, no findings file.
- **`per_pass` is where I could not find prior art.** Per rule 5's uncertain→novel default,
  treat it as novel until someone finds the paper.

**Verify the Huginn detail against the actual paper before anything is stamped novel** — I
am working from recall, not from a fetched source.

## Experiment

`experiments/depth/ablation_harness.py` — `train_one` constructs `CausalRefiner` from plain
kwargs, so `state_noise` threads straight through. Toy scale, CPU, minutes.

Matched pair per rule 2: same seed, same task, same data order, **one knob**.

```
arm A   state_noise=0.0                (control)
arm B   state_noise=σ, mode=init
arm C   state_noise=σ, mode=per_pass
```

σ swept over ~{0.01, 0.03, 0.1} relative to RMS(z). Multi-seed — a single-seed delta is not
a result.

### Pre-registered criteria

Noise floor is σ≈0.03 nats val CE (#17), so the 2σ bar is 0.06.

**Primary readout — depth sensitivity, not loss.** Fraction of positions where argmax
differs between depth 2 and depth 8 on a held-out batch.

- **KEEP** if an arm raises the d2-vs-d8 argmax-disagreement rate by ≥2× over control
  **and** val CE does not regress by more than 0.06 nats.
- **KILL** if depth sensitivity is unchanged, or val CE regresses beyond the bar.
- Loss improving on its own is *not* sufficient to keep — that would be an ordinary
  regulariser, and see the risk below.

## Risks

- **Wrong regime.** Noise regularisers pay off when data-limited across many epochs. We are
  compute-limited on a single pass, where they typically just slow convergence. This is the
  most likely way the idea dies, and the toy ablation is cheap precisely so we find out fast.
- **σ is a new hyperparameter** on an architecture that already has enough of them.
- Toy-scale depth behaviour has transferred imperfectly before — see the depth-transfer
  caveats. A toy KEEP licenses a small GPU run, not a headline.

## Not the same as "creativity"

The original framing paired this with learning creativity. Dropping that: generation
diversity comes from sampling the token distribution (temperature / top-p), which we already
control directly. Train-time latent noise is a regulariser, and at inference we run clean.
The idea stands on the depth argument alone.

## Related follow-up (not this issue)

The sharper diagnostic for whether refinement is *aimed* or *jitter*: project the
depth-induced `Δlogits` onto the top-2 separating direction `e_top1 − e_top2`. Aligned means
refinement works on the decision; orthogonal means it flips near-ties by luck. One forward
pass over a few hundred positions. That measurement would tell us what σ should even be
trying to fix, and arguably should run first.

## Lane

Touches `trm/model/refiner.py`. Toy ablation is CPU-complete and decisive as a gate; a
confirming run needs the card.

**Comment by MatthewLacerda2** [comment]
## Correction — the "sensitivity is emerging" claim in the body is weaker than stated

The run finished; the 100% transcript entry is in. Adding the fourth column changes the read:

| identical pairs (of 8) | 14,400 (47%) | 24,384 (80%) | 27,456 (90%) | **30,464 (100%)** |
|---|---|---|---|---|
| d1 vs d2 | 4/8 | 4/8 | 4/8 | **4/8** |
| d2 vs d4 | 7/8 | 6/8 | 5/8 | **5/8** |
| d4 vs d8 | 8/8 | 7/8 | 6/8 | **7/8** |

`d4 vs d8` went 8 → 7 → 6 → **7**. It did not keep falling. Only `d2 vs d4` moved
consistently (7→6→5→5), and even that is 8 prompts — the binomial error on these counts is
much wider than the movement.

**So the body's "depth sensitivity emerges on its own over training" is over-read from three
points, and I withdraw it as stated.** What survives is narrower: depth 2→4 appears to matter
somewhat more late in training than early; depth 4→8 is indistinguishable from flat.

**This makes the issue's motivation stronger, not weaker.** The saturation between depth 4
and depth 8 is not a transient that training was going to fix on its own — it persisted to
the end of a full 4B-token run. That is the condition this idea is aimed at.

Two knock-ons for whoever picks this up:

- The pre-registered readout (d2-vs-d8 argmax disagreement on a held-out batch) is the right
  measurement and this correction is a reminder of why it is specified that way. **Eight
  prompts of generated text is not it.** Run it over enough positions to have an error bar.
- The related follow-up in the body — projecting depth-induced `Δlogits` onto the top-2
  separating direction — moves up in priority. It gives a per-position, continuous signal
  instead of a rare discrete flip, so it needs far fewer samples to say something.

**Comment by MatthewLacerda2** [comment]
## Premise correction — the motivation in the body above is measurably wrong

Leaving this open, but the reasoning I attached to it has been falsified and the body should
not be read as written.

**What I claimed:** that the refine block "has learned to reach a fixed point almost
immediately and then idle", and that noise was needed to stop it learning a lazy identity.

**What is actually true**, measured on the champion across 2M held-out positions per domain
(#226):

- The block does **not** idle. Depth changes the predicted token on ~1% of positions —
  138σ from zero between depth 1 and depth 4. It moves, and it moves the decision.
- The block also does not **pay**. Paired per-position CE across depths spans ~0.001 nats.
- On `codeparrot` the ordering is coherent (more depth → lower CE); on `fineweb-edu` and
  `finemath` it is scrambled, with depth 1 *beating* depth 4 on both.

So the disease is not inactivity. It is **motion that carries no useful information** — and,
on two of three domains, motion that is slightly counterproductive.

## Does the idea survive its dead motivation?

I think yes, and that is why I am not closing it — but the case has to be rebuilt honestly,
because "stop the block idling" was doing all the work and that job does not exist.

What might still hold: the block currently moves the state without improving the answer. An
objective that forces it to *reconstruct* something specific gives the motion a target it
currently lacks. That is a different argument from the one in the body — it is about
**directing** the work rather than **causing** it — and it is weaker, because nothing here
shows that a denoising target would aim the motion any better than cross-entropy already
does.

## What should gate this

**#227 should run first.** It distinguishes "the loop is aimed at the decision but the readout
cannot see it" from "the loop moves at random." Those imply opposite responses:

- If the motion is **aimed** and the readout wastes it → the fix is the readout (norm
  placement, an untied head), and noise addresses nothing.
- If the motion is **unaimed** → giving it a target is exactly the right shape of fix, and
  this issue gets a real motivation for the first time.

Running this before #227 would be spending a matched pair to answer a question we have a
cheaper instrument for.

## Also revised

The body's "depth sensitivity is emerging on its own over training" was already withdrawn
[in an earlier comment](#issuecomment-5411345976) as over-read from three points. The
in-domain ladder now underway may replace it with a real trend — early indications are that
at 0.78B and 1.40B *more depth is worse*, which would make the picture "depth usefulness
appears late, if at all" rather than "emerges steadily."

Keeping open per the user, to be picked up later. The experimental design in the body (matched
pair, toy scale, depth-sensitivity readout) is still sound — it is only the motivation section
that is dead.

**Comment by MatthewLacerda2** [comment]
superseded-by the depth-recurrence kill (2026-09-12): this concerns the refine loop, which is no longer the live architecture. Record: docs/findings/2026-09-12-depth-recurrence-is-suppressed-not-exploited.md and the ROADMAP graveyard. Re-file against a live mechanism if it ever applies to one.
